### PR TITLE
Add GEP-3155: Complete Backend mTLS Configuration

### PR DESCRIPTION
**Pull request by mkosieradzki**
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
Proposal for backend mTLS configuration

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: geps/gep-3155/index.md
Comment (mikemorris) [inline geps/gep-3155/index.md:98]: This doesn't quite seem accurate, as it seems a client certificate could be presented still if GatewayBackendTLS were configured?

I'm also not quite clear on the use case for allowing this to be specified in two different places, could you elaborate on that in the GEP please?

File: geps/gep-3155/index.md
Comment (mikemorris) [inline geps/gep-3155/index.md:135]:
```suggestion
  // mutually exclusive with URI. At least one of the fields must be specified.
```

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:1]: I know and _most_ people reading this also know that "mTLS" here means "mutual TLS between parties", _without_ the additional Service Mesh concepts that make "mTLS" mean "mutual TLS between services with automatically-managed, rapidly rotated certificates".

But some people, when they read "mTLS", read the latter, not the former.

Could we please change this to "mutual TLS" throughout the document, and put a caveat at the top that "mutual TLS" in this document is used solely to refer to the two-way handshake between Gateway and backend, and _no other mutual TLS use cases_?

@mkosieradzki, I know this may seem pedantic and stupid, but in almost every conversation we've had about client certificate usage, _someone_ has been confused by the use of "mTLS". I just want to avoid having the same issue here as well.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:109]: This limitation is why we did not include client certificate configuration in the initial version of BackendTLSPolicy. Similarly to @mikemorris' comment above, I'd like some use cases that illustrate the need for this, as I think that it's far more likely to be used incorrectly than only having the TLS config for a Gateway present on that Gateway.

To be honest, the BackendTLSPolicy is supposed to be owned by the same party as the Service, and I question if having the party owning the server side of the transaction also specify the client keypair to be used is really fulfilling the spirit of what mutual TLS is supposed to be about.

To put this another way - if you lock your front door, but leave the key tied with a string to the door, are you increasing or decreasing security? Part of the point of adding the second factor of client authentication to a mutual TLS transaction is to ensure that there's some checks and balances in who gets a valid client certificate, and this feels like it bypasses that.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:120]: Again, I think we may need some examples or use cases here to illustrate how this would work, as I _think_ I get it, but am not sure.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:141]: I think that this struct may be better handled as either a type/value struct, as follows:
```suggestion
type SubjectAltName struct {
  // Type chooses the type of SAN stored in this struct. Valid values are either `hostname`
  // or `uri`.
  Type string `json:"type"`
  
  // Value holds the value of the SAN for the type.
  Value string `json:"value"`
}
```

or as a type-selected union, like this:

```suggestion
type SubjectAltName struct {
  // Type chooses the type of SAN stored in this struct. Valid values are either `hostname`
  // or `uri`. The corresponding field must be filled out based on the type (the
  // Hostname or URI fields respectively). If a field not corresponding to the Type is
  // nonzero, it will be ignored.
  Type string `json:"type"`
  
  // Hostname contains Subject Alternative Name specified in DNS name format. It is
  // mutually exclusive with Hostname. At least one of the fields must be specified.
  Hostname v1.PreciseHostname `json:"hostname"`

  // URI contains Subject Alternative Name specified in URI format. It is mutually
  // exclusive with Hostname. At least one of the fields must be specified.
  URI string `json:"uri,omitempty"`
}
```

The first has the advantage of brevity, at the cost of some type-safety and the associated validation on ingestion - it would need more checking in-code instead. The second keeps the type-safety, and works the same as many other type-selected union fields in Kubernetes.

I probably slightly favor the second.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:157]: If there is no SNI, how can the server choose what certificate to serve? This implies but does not state that the server must only have a single serving certificate to choose from. If that's an assumption, I guess that's okay, but we need to be clear about the contract here.

Maybe we need to add something to some docstrings somewhere to say something like "BackendTLSPolicy expects that backends covered by it SHOULD have only one serving keypair to choose from, as the SNI field may not always be set" or something.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:184]: I understand that these escape hatches are handy, but as a maintainer I'm concerned that they have two effects:

* They promote the creation of non-portable configuration, which is against the purpose of this API
* They encourage users to use them and not go through the (yes, sometimes arduous) process of giving feedback and building features into upstream.

The fact that you're saying this is useful and I have no idea how this field is actually being used in the field illustrates exactly what I mean. If you can get everything you need out of the spec by using a free-text escape hatch like this, why would you ever do the work to come and ask for proper typed, structured fields?

To put this another way: everytime I see a `map[string]string`, I think of Ingress annotations and the terrible situation that has ended up in. So I, personally, need to see a pretty big benefit to justify these risks. I'd like to see some concrete use cases here as well that illustrate the need for this.

**Review by youngnick — CHANGES_REQUESTED** [review]
A few pieces of feedback here, but I think that the number one thing that's needed is use cases that show how some of these proposals work, or justify their existence.

In particular, I'm reluctant to add free-text `map[string]string` fields to the API without a _very_ good explanation and example use cases.

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:98]: The idea here is to allow overriding the default from GatewayBackendTLS with no certificate.

This is to support some compatibility corner cases, for example around a pre-packaged software.

Let's imagine a prepackaged software that always requests a certificate, but if customer refuses, it still proceeds with connection but falls back to an alternative authentication method, e.g. JWT. But for some reason if it observes an unexpected certificate it rejects the connection.

If we always presented the certificate in such case, it would be impossible to use this prepackaged software correctly, without creating a dedicated gateway with TLS disabled.

Having said that, if there is a consensus that this corner case is too far fetched, I am happy to drop it.

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:1]: Thank you, I was not aware of that. I will update the doc accordingly to avoid confusion with managed TLS or managed mutual TLS.

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md]: Thanks, I will update to type-selected union.

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:157]: I think that the common convention is to provide the "default" certificate in such case.

I think I would rather use language that the Hostname SHOULD be sad - otherwise server will have hard time selecting the correct certificate, but I wanted to allow unset for compatibility corner cases.

But again, if there is a consensus we don't need to support this case -> I am more than happy to drop this from the spec.

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:184]: I believe this will be primarily used for experimental features, before they are actually merged. First two I expect are:
- constraints like min-tls-version
- configuration for proprietary / experimental managed TLS modes

File: geps/gep-3155/index.md
Comment (mikemorris) [reply geps/gep-3155/index.md:98]: Expecting an undefined or explicit `nil` here to override a configured value feels awkward and potentially confusing. It feels like the right way to handle this need would be to somehow scope GatewayBackendTLS better to only be applied for some backend connections?

File: geps/gep-3155/index.md
Comment (gcs278) [inline geps/gep-3155/index.md:1]: Just a reminder, you'll need to also update `mkdocs.yml` so that it shows in https://gateway-api.sigs.k8s.io/geps/overview/ as well as  `geps/gep-3155/metadata.yaml`

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:109]: Added the use cases, added the write-up on security of this solution, can you PTAL at the updated doc?

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:98]: Agreed, replaced it with an Enum

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:157]: Agreed, this seems to be pretty niche, let's remove it.

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:1]: Thanks, done

File: geps/gep-3155/index.md
Comment (howardjohn) [inline geps/gep-3155/index.md:120]: Does this mean I can have:
* Gateway in ns1, selecting route
* Route in ns2, targeting svc in ns3
* BackendTLSPolicy is ns3 targetting svc (no cross namespace allowed here), pointing to a cert in ns4
* Cert in ns4 with refgrant allowing backendTLS policy

So ultimately I have a chain of 4 namespaces allowing the Gateway is ns1 to read a secret in ns4?

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:66]:
```suggestion
  // Certificate and the associated private key.
```

I think we need to be clear that this needs to be a full keypair, not just the certificate.

File: geps/gep-3155/index.md
Comment (youngnick) [reply geps/gep-3155/index.md:120]: That seems valid from what I can see, and is a good example of how quickly complexity can build  out of small changes.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:83]: I don't know the best way to word this, but it really feels like we need to tell people that you shouldn't use this without a good reason. Because otherwise people will look at the spec of `BackendTLSPolicy` and assume that this is the right place to configure client certificate details.

I also still think that having this config makes the client certificate used here basically worthless for actual authentication - it is essentially `curl -k` for Gateways.

As it is, I suspect that most examples we see will use this and not the Gateway level config, because this is easily within the control of the service owner, and requires no interaction with the Gateway owner.

These feelings aside, the actionable feedback here is to have some sort of callout that Client Certificate config should preferably be in the Gateway stanza, not this one. This should only be used when the Gateway option doesn't work out for some reason.

File: geps/gep-3155/index.md
Comment (youngnick) [reply geps/gep-3155/index.md:160]: Maybe I should be clearer here. Currently, what this field is actually used for is not defined at all. Being able to specify more SANs is great, but as an implementer, what do I do with this field?

From the below struct docs, it seems like the answer is maybe "The serving certificate received when the Gateway implementation makes a client TLS connection to the backend must have at least one of the specified SANs set".

This is _critical_, because if we go ahead with also adding client certificates to this object, we now have both client and serving certificates on the same object. So it's really, really important that we leave no wiggle room here in the godoc.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:213]: So, `Hostname` still sets the SNI that is sent as part of the initial request, but now no longer sets anything about how the server certificate must be checked?

In that case, I'd recommend something like this:
```suggestion
  // 2. Hostname MUST be used for authentication and MUST match the certificate
  //     served by the matching backend UNLESS SubjectAltNames is specified. If
  //     SubjectAltNames is specified, Hostname MUST only be used for setting
  //     the SNI sent, and the SubjectAltNames MUST be used for validating the
  //     serving certificate.
```

It's saying the same thing, but it is leaving less room for misunderstanding.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:225]: I just want to be on record as saying that I didn't like this field when we added it to Gateway, I don't like this field, but I don't have a good enough reason to block any changes. Let's see if it ends up misused in the wild.

**Review by youngnick — COMMENTED** [review]
Smaller changes, but I don't have any blockers once we find a way to finish up these last things.

**Review by youngnick — COMMENTED** [review]
Smaller changes, but I don't have any blockers once we find a way to finish up these last things.

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:120]: Yes, this seems to be a valid scenario, do you perceive this as a problem?

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md]: Thanks a lot!

File: geps/gep-3155/index.md
Comment (howardjohn) [reply geps/gep-3155/index.md:120]: Maybe? The principles as I understand it is there should be a strong two way handshake between each cross namespace hop.

So the question we have is did the `ns4` owner authorization `ns1` to read its Secrets.

Working backwards:
* ns4 granted ns3 Secret read access explicitly with RefGrant. :heavy_check_mark: 
* ns3 explicitly granted ns2 access to **send to services**. :stop_sign: 
* ns1 explicitly granted ns2 access to bind routes :heavy_check_mark: 

I feel there is a violation here. There is no strong agreement from ns3 that ns2 can access its Secrets -- only that it can access its services.

Comment (mkosieradzki) [reply geps/gep-3155/index.md:120]: I agree with this concern - this calls at least for a clear warning when:
- BackendTLSPolicy is ns3 targetting svc (no cross namespace allowed here), pointing to a cert in ns4

ns3 owner should be aware that it is exposing the cert to all gateways that are allowed to **send to this service** will be able to access the certificate

Comment (howardjohn) [reply geps/gep-3155/index.md:120]: I don't really think that is sufficient. Today, there may be many many users giving Service access. This is an **extremely low** privilege to grant, so users will hand it out freely.

Now we are saying that in a new release, that the persona is not even aware of, suddenly granting Service (low priv) gives access to Secret (high priv)

Comment (mkosieradzki) [reply geps/gep-3155/index.md:120]: How about requiring two grants, i.e. ?
- ns4 to ns3 secret (using secret to connect to the backend you control)
- ns1 to ns3 secret (to consume the secret)

File: geps/gep-3155/index.md
Comment (robscott) [inline geps/gep-3155/index.md:218]:
```suggestion
### Allow per-service TLS settings BackendTLSPolicy
```

File: geps/gep-3155/index.md
Comment (robscott) [reply geps/gep-3155/index.md:227]: I definitely agree with the hesitation here, but ultimately believe this is more helpful than not. There's such a wide range of TLS config possible that it seems unlikely that we'll ever cover all of the potential concepts in the core API. I think it's likely preferable to mirror the options map we have on Gateway Listeners than to leave out a form of extensibility here, though I'll admit that both options have risks.

File: geps/gep-3155/index.md
Comment (robscott) [reply geps/gep-3155/index.md:160]: +1, good points @youngnick. I think the godoc below describes the behavior fairly clearly, maybe it would be helpful to pull that up to this paragraph as well?

> When specified, at least one of certificate's Subject Alternate Names MUST match at least one of the specified SubjectAltNames.

File: geps/gep-3155/index.md
Comment (robscott) [inline geps/gep-3155/index.md:159]: Should hostname still be a required field and used for SNI? Or are you suggesting that it's optional and SNI can be _anything_ when SANs are specified?

File: geps/gep-3155/index.md
Comment (robscott) [reply geps/gep-3155/index.md:120]: This is a good point. A few options I can think of:

**1. References to Secrets that are used by the Gateway are relative to that Gateway**
With this option, we would consider the reference to this Secret as coming from the Gateway when it comes to cross-namespace logic. 

That means that a reference from BackendTLSPolicy in ns3 to a cert in ns1 would be considered _local_ and not need a ReferenceGrant, because the _Gateway_ is in ns1. On the other hand, a reference from BackendTLSPolicy in ns3 to a Secret in ns3 would require a ReferenceGrant that allowed a reference from the Gateway in ns1 to the secret in ns3.

**2. References are relative to BackendTLSPolicy, BackendTLSPolicy should be considered a privileged resource**

I've increasingly been making the case that Service owners aren't necessarily the correct owners for BackendTLSPolicy, particularly in larger organizations. If we consider BackendTLSPolicy a privileged resource, I think it's reasonable to say that valid references from this resource inherently grant access to the Gateway(s)  implementing that resource.

**3. Namespace-local for now**
I don't think we have clear evidence that we _need_ cross-namespace references at this level. Given the complexity involved, it may be safer to explicitly prohibit that as a starting point. We could instead require the Gateway, BackendTLSPolicy, and cert to all be in the same namespace when certs are specified at this level. That does admittedly feel rather onerous, but it's easier to add cross-namespace support in the future than to retroactively fix things if we leave things to open-ended to start with.


With all that said, I'm leaning towards the third option here. Open to other perspectives.

**Review by robscott — COMMENTED** [review]
Thanks @mkosieradzki!

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:120]: I like the idea of requiring BackendTLSPolicy to be in the very same namespace as Gateway, but I am afraid this ship has sailed...  as per: https://gateway-api.sigs.k8s.io/api-types/backendtlspolicy/
"The BackendTLSPolicy and the Service must reside in the same namespace in order to prevent the complications involved with sharing trust across namespace boundaries."

Comment (mkosieradzki) [reply geps/gep-3155/index.md:120]: I believe we should park this discussion until we reach conclusion #3226, I am moving this overrides part to the future work.

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:54]: Added this sentence, does it look better now?

File: geps/gep-3155/index.md
Comment (mkosieradzki) [reply geps/gep-3155/index.md:53]: It cannot by anything, because it needs to be deterministic what we send to the backend. So yes hostname is still required.

File: geps/gep-3155/index.md
Comment (jaishals) [inline geps/gep-3155/index.md:1]: Hello @mkosieradzki , if I understood the last iterations of GEP correctly, we are proposing that the BackendTLSPolicy spec target reference will be only gateway and not service ? If that is true , how would the scenarios where there are mix of services which do not need MTLS coexist in the same Gateway -> Route hierarchy.

File: geps/gep-3155/index.md
Comment (mikemorris) [inline geps/gep-3155/index.md:43]: Re `all` backends - should the client cert be presented even when no BackendTLSPolicy targeting the Service backend is present?

File: geps/gep-3155/index.md
Comment (mikemorris) [inline geps/gep-3155/index.md:47]: I think this is intended to be applicable to the "server"/destination cert, but can you please specify more precisely which cert is affected here?

File: geps/gep-3155/index.md
Comment (mikemorris) [inline geps/gep-3155/index.md:65]: What happens if it doesn't? Suggested `status` condition on the policy indicated it's invalid?

File: geps/gep-3155/index.md
Comment (mikemorris) [inline geps/gep-3155/index.md:41]: This is missing an API example - seems to be suggesting the addition of a new field under the Gateway `spec` stanza https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.Gateway?

**Review by mikemorris — CHANGES_REQUESTED** [review]

File: geps/gep-3155/index.md
Comment (robscott) [reply geps/gep-3155/index.md:65]: Apparently we haven't defined _how_ a request should be rejected here:

https://github.com/kubernetes-sigs/gateway-api/blob/c94d4010a1524e86ffd07bc5ccedde340dba5843/apis/v1alpha3/backendtlspolicy_types.go#L128-L130

At a minimum I think we can agree that the request should be rejected, ideally we can be more specific here for the sake of writing conformance tests and overall portability.

As far as populating status, I think that most controllers are separated from the dataplane and would not be able to translate runtime TLS validation errors to config time (k8s) status.

**Review by robscott — COMMENTED** [review]
Thanks @mkosieradzki! A few more comments, but mostly LGTM.

File: geps/gep-3155/index.md
Comment (mikemorris) [reply geps/gep-3155/index.md:43]: > Ideally Gateway owners could provide a set of CAs they trust for certs from backends at the Gateway level

I think this makes sense and could potentially address the non-mutual end-to-end TLS functionality @jaishals is describing in https://github.com/kubernetes-sigs/gateway-api/pull/3180#discussion_r1697382336, where the presence of BackendTLSPolicy additional validation is specifically to enforce _mutual_ TLS? Would any additional signal like `mode: mutual` be needed somewhere to clarify this?

File: geps/gep-3155/index.md
Comment (youngnick) [reply geps/gep-3155/index.md:54]: Sorry, but I think we really need the mention that it's the _serving_ cert that must be matched against.

So I think that sentence should be:

`When specified, the certificate served from the backend MUST have at least one Subject Alternate Name matching one of the specified SubjectAltNames`.

File: geps/gep-3155/index.md
Comment (youngnick) [inline geps/gep-3155/index.md:65]: Similarly to above, I think this should be:
```suggestion
  // When specified, the certificate served by the backend MUST have at a Subject Alternate Name
  // that matches one of the specified SubjectAltNames.
```

Just so it's _extremely_ clear what certificate this is talking about.

File: geps/gep-3155/index.md
Comment (youngnick) [reply geps/gep-3155/index.md:41]: Yes, I think this should have a quick sketch of what we need to add to Gateway.

**Review by youngnick — COMMENTED** [review]
I've been picky about wording, but once the two wording changes I've asked for are handled, and we have an example of what we are adding to Gateway, then this LGTM.

File: geps/gep-3155/index.md
Comment (robscott) [inline geps/gep-3155/index.md:177]: Our docs gen requires a newline before a list starts (https://deploy-preview-3180--kubernetes-sigs-gateway-api.netlify.app/geps/gep-3155/#conformance-details)

```suggestion
1. When SubjectAltNames are specified in BackendTLSPolicy:

  - The hostname field is still used as SNI, if specified
```

File: geps/gep-3155/index.md
Comment (robscott) [inline geps/gep-3155/index.md:182]:
```suggestion
2. When a Client Certificate is specified on a Gateway:

  - It is applied to all services.
```

**Review by robscott — APPROVED** [review]
Thanks @mkosieradzki! Added a few formatting nits but otherwise LGTM.

/approve

**Review by youngnick — APPROVED** [review]

**Comment by youngnick** [comment]
With that last round of changes, this LGTM. I'll leave the final unhold for @robscott though, since he has some outstanding format changes.

/lgtm

**Comment by robscott** [comment]
Thanks @mkosieradzki!

/lgtm
/hold cancel

**Pull request merged**
